tooling/image-sync/repository.go

```diff
@@ -0,0 +1,363 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/containers/azcontainerregistry"
+)
+
+// Registry is the interface for accessing image repositories
+type Registry interface {
+	GetTags(context.Context, string) ([]string, error)
+}
+
+// AuthedTransport is a http.RoundTripper that adds an Authorization header
+type AuthedTransport struct {
+	Key     string
+	Wrapped http.RoundTripper
+}
+
+// RoundTrip implements http.RoundTripper and sets Authorization header
+func (t *AuthedTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Set("Authorization", t.Key)
+	return t.Wrapped.RoundTrip(req)
+}
+
+// QuayRegistry implements Quay Repository access
+type QuayRegistry struct {
+	httpclient   *http.Client
+	baseUrl      string
+	numberOftags int
+}
+
+// NewQuayRegistry creates a new QuayRegistry access client
+func NewQuayRegistry(cfg *SyncConfig, bearerToken string) *QuayRegistry {
+	return &QuayRegistry{
+		httpclient: &http.Client{Timeout: time.Duration(cfg.RequestTimeout) * time.Second,
+			Transport: &AuthedTransport{
+				Key:     "Bearer " + bearerToken,
+				Wrapped: http.DefaultTransport,
+			},
+		},
+		baseUrl:      "https://quay.io",
+		numberOftags: cfg.NumberOfTags,
+	}
+}
+
+type TagsResponse struct {
+	Tags          []Tags
+	Page          int
+	HasAdditional bool
+}
+
+type Tags struct {
+	Name string
+}
+
+// GetTags returns the tags for the given image
+func (q *QuayRegistry) GetTags(ctx context.Context, image string) ([]string, error) {
+	// Todo pagination
+	Log().Debugw("Getting tags for image", "image", image)
+	path := fmt.Sprintf("%s/api/v1/repository/%s/tag/", q.baseUrl, image)
+	req, err := http.NewRequestWithContext(ctx, "GET", path, nil)
+
+	Log().Debugw("Sending request", "path", path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %v", err)
+	}
+	resp, err := q.httpclient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send request: %v", err)
+	}
+	Log().Debugw("Got response", "statuscode", resp.StatusCode)
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code %d", resp.StatusCode)
+	}
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response: %v", err)
+	}
+
+	var tagsResponse TagsResponse
+	err = json.Unmarshal(body, &tagsResponse)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal response: %v", err)
+	}
+
+	var tags []string
+
+	for _, tag := range tagsResponse.Tags {
+		if tag.Name == "latest" {
+			continue
+		}
+		tags = append(tags, tag.Name)
+		if len(tags) >= q.numberOftags {
```

Comment:
the quay API endpoint to list tags supports a paramter `limit`
https://docs.quay.io/api/swagger/#!/tag/listRepoTags

using this would make `ReadAll` and `Unmarshal` run faster as well.

Reply:
also if we are interested in more than 100 images, we need paging as the max value for `limit` is 100

Reply:
I wanted to overhaul this later, but I can also do it now to avoid broken/unfinished code.

Reply:
I Added pagination, but still keeping the limit of 100 here. The reason is, that _latest_ can occur a couple of times in the response. The API will return one entry for _latest_, every time it is pushed. Probably page size could be `numberOftags * 2`. But adding this complexity is not worth vs. having todo extra requests. Which will undoubtedly be slower than reading a couple of more bytes.